storage/integrate.go

```diff
@@ -0,0 +1,337 @@
+		return nil, err
+	}
+	return t.rf.NewRange(0, treeSize, hashes)
+}
+
+func (t *TreeBuilder) Integrate(ctx context.Context, fromSize uint64, entries [][]byte) (newSize uint64, rootHash []byte, tiles map[TileID]*api.HashTile, err error) {
+	baseRange, err := t.newRange(ctx, fromSize)
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("failed to create range covering existing log: %w", err)
+	}
+
+	// Initialise a compact range representation, and verify the stored state.
+	r, err := baseRange.GetRootHash(nil)
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("invalid log state, unable to recalculate root: %w", err)
+	}
+
+	klog.V(1).Infof("Loaded state with roothash %x", r)
+	// Create a new compact range which represents the update to the tree
+	newRange := t.rf.NewEmptyRange(fromSize)
+	if len(entries) == 0 {
+		klog.V(1).Infof("Nothing to do.")
+		// Nothing to do, nothing done.
+		return fromSize, r, nil, nil
+	}
+	tc := newTileWriteCache(fromSize, t.getTile)
+	visitor := tc.Visitor(ctx)
+	for _, e := range entries {
+		lh := rfc6962.DefaultHasher.HashLeaf(e)
+		// Update range and set nodes
+		if err := newRange.Append(lh, visitor); err != nil {
+			return 0, nil, nil, fmt.Errorf("newRange.Append(): %v", err)
+		}
+
+	}
+
+	// Merge the update range into the old tree
+	if err := baseRange.AppendRange(newRange, visitor); err != nil {
+		return 0, nil, nil, fmt.Errorf("failed to merge new range onto existing log: %w", err)
+	}
+
+	// Calculate the new root hash - don't pass in the tileCache visitor here since
+	// this will construct any ephemeral nodes and we do not want to store those.
+	newRoot, err := baseRange.GetRootHash(nil)
+	if err != nil {
+		return 0, nil, nil, fmt.Errorf("failed to calculate new root hash: %w", err)
+	}
+
+	if err := tc.Err(); err != nil {
+		return 0, nil, nil, err
+	}
+
+	// All calculation is now complete, all that remains is to store the new
+	// tiles and updated log state.
+	klog.V(1).Infof("New log state: size 0x%x hash: %x", baseRange.End(), newRoot)
+
+	return baseRange.End(), newRoot, tc.Tiles(), nil
+
+}
+
+// tileReadCache is a structure which provides a very simple thread-safe map of tiles.
+type tileReadCache struct {
+	sync.RWMutex
+
+	entries map[string]*populatedTile
+}
+
+func newTileReadCache() tileReadCache {
+	return tileReadCache{
+		entries: make(map[string]*populatedTile),
+	}
+}
+
+// Get returns a previously set tile and true, or, if no such tile is in the cache, returns nil and false.
+func (r *tileReadCache) Get(tileID TileID, treeSize uint64) (*populatedTile, bool) {
+	r.RLock()
+	defer r.RUnlock()
+	k := layout.TilePath(uint64(tileID.Level), tileID.Index, treeSize)
+	e, ok := r.entries[k]
+	return e, ok
+}
+
+// Set associates the given tileID with a tile.
+func (r *tileReadCache) Set(tileID TileID, treeSize uint64, t *populatedTile) {
+	r.Lock()
+	defer r.Unlock()
+	k := layout.TilePath(uint64(tileID.Level), tileID.Index, treeSize)
+	if e, ok := r.entries[k]; ok && !e.Equals(t) {
```

Comment:
You should be able to drop this equality check if the read-once semantics for any given tile are fixed.

Reply:
That's gone away now